### PR TITLE
Gravatar: Implement passwordless login flow

### PR DESCRIPTION
**Pull request by wellyshen**
Related to 105630-gh-Automattic/gravatar

## Proposed Changes

* Customize the new Gravatar login page
* Customize the new Gravatar magic login page and enable new account creation
* Customize the new Gravatar email verification page and implement the resend email feature
* Clear the old Gravatar login page-related code
* Fix the wrong signup URLs on the related F2A pages
* Fix an existing lint error by using `localizeUrl` for the "Get help" link for F2A pages

## Testing Instructions

1. Setup Calypso repo by following [this doc](https://github.com/Automattic/wp-calypso#getting-started) and check out this PR
2. Run `yarn install` and then `yarn start`
3. Logout your account, and open [this magic login page](http://calypso.localhost:3000/log-in/link?redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%3Fclient_id%3D1854%26response_type%3Dcode%26blog_id%3D0%26state%3D111ffb2efc6d68ee7a6379a556cc6a9195c2ad255a2f310f7042213fd5c4ee29%26redirect_uri%3Dhttps%253A%252F%252Fen.gravatar.com%252Fconnect%252F%253Faction%253Drequest_access_token%26from-calypso%3D1&client_id=1854)
4. Try to **signup a new account** (you can use [Gmail testing accounts](https://www.google.com/search?q=gmail+test+accounts&rlz=1C5CHFA_enTW1017TW1017&oq=gmail+test+accounts&gs_lcrp=EgZjaHJvbWUqBggAEEUYOzIGCAAQRRg7MggIARAAGBYYHjIICAIQABgWGB4yCggDEAAYhgMYigUyCggEEAAYhgMYigUyCggFEAAYhgMYigUyBggGEEUYPDIGCAcQRRg80gEINTAzNWowajeoAgCwAgA&sourceid=chrome&ie=UTF-8)), and the flow should work as below:

https://github.com/Automattic/wp-calypso/assets/21308003/668349e8-027a-42e7-9af5-ab3d23fecf30

5.  Logout your account, and try to **login with a non-A11n + F2A account** via [this magic login page](http://calypso.localhost:3000/log-in/link?redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%3Fclient_id%3D1854%26response_type%3Dcode%26blog_id%3D0%26state%3D111ffb2efc6d68ee7a6379a556cc6a9195c2ad255a2f310f7042213fd5c4ee29%26redirect_uri%3Dhttps%253A%252F%252Fen.gravatar.com%252Fconnect%252F%253Faction%253Drequest_access_token%26from-calypso%3D1&client_id=1854), the flow should work as below:

https://github.com/Automattic/wp-calypso/assets/21308003/d9d3e2b4-f63e-4f23-b209-222c1031c360

6. Check the UIs and click the following links, they should work as expected:

| Magic login | Check your email |
| - | - |
| 截圖 2023-09-01 下午3 11 08 | 截圖 2023-09-01 下午3 11 29 |

7.  Logout your account, and try to **login with a passwordless account (like we did via No.5)** via [this login page](http://calypso.localhost:3000/log-in?redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%3Fclient_id%3D1854%26response_type%3Dcode%26blog_id%3D0%26state%3D111ffb2efc6d68ee7a6379a556cc6a9195c2ad255a2f310f7042213fd5c4ee29%26redirect_uri%3Dhttps%253A%252F%252Fen.gravatar.com%252Fconnect%252F%253Faction%253Drequest_access_token%26from-calypso%3D1&client_id=1854), and the flow should work as below:

https://github.com/Automattic/wp-calypso/assets/21308003/840802f4-616b-4a6f-9e17-ba62cfe6e9af

8. Check the UIs and click the following links, they should work as expected:

截圖 2023-09-01 下午3 46 49

9. Login with an F2A account via [this login page](http://calypso.localhost:3000/log-in?redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%3Fclient_id%3D1854%26response_type%3Dcode%26blog_id%3D0%26state%3D111ffb2efc6d68ee7a6379a556cc6a9195c2ad255a2f310f7042213fd5c4ee29%26redirect_uri%3Dhttps%253A%252F%252Fen.gravatar.com%252Fconnect%252F%253Faction%253Drequest_access_token%26from-calypso%3D1&client_id=1854), and click the following signup links, you will be redirect to Gravatar's signup page:

| F2A app page | F2A code page | F2A backup code page |
| - | - | - |
| 截圖 2023-09-01 下午4 12 00 |  截圖 2023-09-01 下午4 12 07 | 截圖 2023-09-01 下午4 12 15 |

10. Check WPCOM's login page, it should work the same:
  - [Dev](http://calypso.localhost:3000/log-in)
  - [Prod](https://wordpress.com/log-in)

11. Check WPCOM's magic login page, it should work the same:
  - [Dev](http://calypso.localhost:3000/log-in/link)
  - [Prod](https://wordpress.com/log-in/link)

12. Check any OAuth2 clients, e.g. Jetpack, it should work the same:
  - [Dev](http://calypso.localhost:3000/log-in?client_id=69040&redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%3Fresponse_type%3Dtoken%26client_id%3D69040%26redirect_uri%3Dhttps%253A%252F%252Fcloud.jetpack.com%252Fconnect%252Foauth%252Ftoken%253Fnext%253D%25252F%26scope%3Dglobal%26blog_id%3D0%26from-calypso%3D1)
  - [Prod](https://wordpress.com/log-in?client_id=69040&redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%3Fresponse_type%3Dtoken%26client_id%3D69040%26redirect_uri%3Dhttps%253A%252F%252Fcloud.jetpack.com%252Fconnect%252Foauth%252Ftoken%253Fnext%253D%25252F%26scope%3Dglobal%26blog_id%3D0%26from-calypso%3D1)

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/login/magic-login/request-login-email-form.jsx [hunk 1]
```diff
@@ -121,9 +138,37 @@ class RequestLoginEmailForm extends Component {
...
+								<a
+									href={ localizeUrl( 'https://automattic.com/privacy/' ) }
+									target="_blank"
+									rel="noopener noreferrer"
+								/>
+							),
+						},
+					}
+				) }
+			</div>
```
Comment (aaronfc) [inline client/login/magic-login/request-login-email-form.jsx:164]: This seems to be center-aligned in the designs in [here](IXT40Xck14laWK456R4wB2-fi-2476%3A20672&mode=dev).

**Review by aaronfc — COMMENTED** [review]
Hey @wellyshen this is looking good and everything I tested worked fine.

Let me say first that I do not have much expertise in front-end development so I would really like some extra feedback on the frontend code from someone who knows better. Maybe from @Automattic/lighthouse ?

A part from that (and the couple of comments made on the code), I did some tests and these are my findings:
- 🔴 Login flow ends up in WordPress.com (instead of Gravatar) if you are **already** logged in WordPress.com. To reproduce: Login to WordPress.com and try to complete the passwordless login flow. You will be redirected to WordPress.com instead of Gravatar.
- 🟡  It was a bit confusing the mixture of WordPress/Gravatar/Autommatic links. This might be expected but it is strange from a user perspective.
  - TOS: wordpress.com
  - Privacy: automattic.com
  - Help docs: gravatar.com
  - Lost password in SSO login page: wordpress.com. In my opinion this one is specially weird. Since out of the blue it takes you to a very different page from WordPress.com.
- 🟡 The "Resend the verification email" worked every time I clicked it. I clicked 5 times in a row and received 5 emails. Maybe we should "disable" it temporarily or add some cooldown?
- 🟡  Related to the same 'resend verification email' flow. The 5 emails I received worked. I would have expected only the **last one** to work. But this might be related to WordPress.com mechanism and I am not sure we can do anything around this now.
- 🔴  2FA behaved strangely in my case. I do have authenticator app and also the passkeys/codes enabled.
  - When I followed the process first it asked the authenticator app code.
  - After adding it I landed again in the login screen (strange!).
  - After completing the login a second time I was now asked the passkey/code.
  - Completed the login with the fingerprint and then I was successfully logged in.
- 🟡  When logging out from Gravatar I was logged out also from WordPress.com. This is maybe expected, but from a user perspective might be strange since there is no indication about that.

A part from the above, @wellyshen I remember you mentioned me that signup flow will be dropped. And only the login flow will be used. I saw there are some scenarios where we point to the signup flow:
- "Create account" link in the Google/Apple login screen.
- "Your account does not exist 'create one'" in the feedback when you add a non-existing email in Google/Apple login screen.
My guess is we should point those to the login flow instead of the signup one?

File: client/blocks/login/index.jsx [hunk 2]
```diff
@@ -554,6 +558,9 @@ class Login extends Component {
 
 		return (
 			<div className="login__form-header-wrapper">
+				{ isGravatarLoginPage && (
+					<img src={ oauth2Client.icon } width={ 27 } height={ 27 } alt={ oauth2Client.alt } />
```
Comment (aaronfc) [inline client/blocks/login/index.jsx]: I am not sure if `oauth2Client.alt` exists. Also you are using `oauth2Client.title` in [here](https://github.com/Automattic/wp-calypso/pull/81258#diff-53db3b94011d088eb9d347bbe08dd60fd909d858000f6f536e190ec93384bbacR145).

Comment (aaronfc) [reply client/blocks/login/index.jsx]: Same [here](https://github.com/Automattic/wp-calypso/pull/81258/files#diff-53db3b94011d088eb9d347bbe08dd60fd909d858000f6f536e190ec93384bbacR211).

Comment (wellyshen) [reply client/blocks/login/index.jsx]: Good catch, I have fixed this issue, thanks!

File: client/login/magic-login/request-login-email-form.jsx (hunk 1 above)
Comment (wellyshen) [reply client/login/magic-login/request-login-email-form.jsx:164]: Good catch! I have fixed this issue, thank you!

**Comment by wellyshen** [comment]
@aaronfc Thank you for the review. The entire login flow consists of many components, some of which are live in the WP.com codebase. However, this PR specifically focuses on the passwordless/magic-link portion of this Figma design: IXT40Xck14laWK456R4wB2-fi. Apologies for not mentioning that initially.

> 🔴 Login flow ends up in WordPress.com (instead of Gravatar) if you are already logged in WordPress.com. To reproduce: Login to WordPress.com and try to complete the passwordless login flow. You will be redirected to WordPress.com instead of Gravatar.

During testing, we directly access the login URL. However, under normal circumstances, logged-in users would not be directed to the flow. Therefore, I'm not worried about this issue at this stage.

> 🟡 It was a bit confusing the mixture of WordPress/Gravatar/Autommatic links. This might be expected but it is strange from a user perspective.
TOS: wordpress.com
Privacy: automattic.com
Help docs: gravatar.com

As of now, Gravatar doesn't have its own specific ToS or Privacy Policy. It's worth noting that the privacy link on Gravatar's site redirects to [Automattic's privacy page](https://automattic.com/privacy/). While it's unclear whether dedicated policies for Gravatar will be implemented in the future, we are currently using Automattic's policies for safety reasons.

截圖 2023-09-06 下午8 06 43

> Lost password in SSO login page: wordpress.com. In my opinion this one is specially weird. Since out of the blue it takes you to a very different page from WordPress.com.

The page is not part of Calypso; it resides in the WordPress.com codebase. I have created a related issue for it: 105690-Automattic/Gravatar. Thanks!

截圖 2023-09-06 下午11 56 03

> 🟡 The "Resend the verification email" worked every time I clicked it. I clicked 5 times in a row and received 5 emails. Maybe we should "disable" it temporarily or add some cooldown?

Thank you for bringing this issue to my attention. I have created the corresponding issue: 105691-Automattic/gravatar.

> 🟡 Related to the same 'resend verification email' flow. The 5 emails I received worked. I would have expected only the last one to work. But this might be related to WordPress.com mechanism and I am not sure we can do anything around this now.

This feature should be implemented on the backend. I'm not certain if we're capable of doing that as well. Nonetheless, I have created a related issue for it: 105692-Automattic/gravatar.

> 🔴 2FA behaved strangely in my case. I do have authenticator app and also the passkeys/codes enabled.
When I followed the process first it asked the authenticator app code.
After adding it I landed again in the login screen (strange!).
After completing the login a second time I was now asked the passkey/code.
Completed the login with the fingerprint and then I was successfully logged in.

I am unable to test this in the development environment, so I will verify it in production. Thank you!
> 🟡 When logging out from Gravatar I was logged out also from WordPress.com. This is maybe expected, but from a user perspective might be strange since there is no indication about that.

If I recall correctly, some users have complained that Gravatar does not log them out of their accounts. This issue arises because Gravatar accounts are linked to WP.com accounts. If we don't log out of WP.com, we won't be able to log out of Gravatar either. We have resolved this issue by ensuring that logging out of WP.com also logs users out of Gravatar. For more details, please refer to this issue: 105265-Automattic/gravatar.

> A part from the above, @wellyshen I remember you mentioned me that signup flow will be dropped. And only the login flow will be used. I saw there are some scenarios where we point to the signup flow:
>
> "Create account" link in the Google/Apple login screen.
"Your account does not exist 'create one'" in the feedback when you add a non-existing email in Google/Apple login screen.
My guess is we should point those to the login flow instead of the signup one?

The signup flow is currently still available for Gravatar users. Once we decide to discontinue this signup flow, we will update the relevant signup links and potentially make some design adjustments through a separate PR.

**Comment by aaronfc** [comment]
Hey, thanks for the detailed responses!

> During testing, we directly access the login URL. However, under normal circumstances, logged-in users would not be directed to the flow. Therefore, I'm not worried about this issue at this stage.

I am not completely sure about this. I understand your point, but since this is an asynchronous process (sending an email). The users could login to WordPress.com before clicking the email link. And then they would land in WP.com.
Having said that, you are right. This might not be directly related to this PR, since I tested and it's already doing the same in production. Maybe is it worth a ticket to look into it?

> The page is not part of Calypso; it resides in the WordPress.com codebase. I have created a related issue for it: 105690-Automattic/Gravatar. Thanks!

Great! Thanks for the explanation!

> Thank you for bringing this issue to my attention. I have created the corresponding issue: 105691-Automattic/gravatar.

Cool!

> This feature should be implemented on the backend. I'm not certain if we're capable of doing that as well. Nonetheless, I have created a related issue for it: 105692-Automattic/gravatar.

🆒 

> I am unable to test this in the development environment, so I will verify it in production. Thank you!

After doing a couple more tests I realised this probably has to do with the passkeys being configured for the production domain (and obviously not working in calypso development environment). I am ok with testing this once in production :)

> If I recall correctly, some users have complained that Gravatar does not log them out of their accounts. This issue arises because Gravatar accounts are linked to WP.com accounts. If we don't log out of WP.com, we won't be able to log out of Gravatar either. We have resolved this issue by ensuring that logging out of WP.com also logs users out of Gravatar. For more details, please refer to this issue: 105265-Automattic/gravatar.

Understood!

> The signup flow is currently still available for Gravatar users. Once we decide to discontinue this signup flow, we will update the relevant signup links and potentially make some design adjustments through a separate PR.

Great! We might need to hide those for WPJMCOM task, but I am not 100% now. Will look into the changes we need (as we discussed through slack) and come back to you.

For the record: We discovered there are duplicated `/authenticate` requests being made because of the way we are testing – we are starting the flow with a `state` that is unknown to `gravatar.com`. This will be fixed in production since we would have created the `state` initially in `gravatar.com` and then redirecting to `/authenticate`. Similar to what I am doing in WPJMCOM [here](626-gh-Automattic/wpjobmanager.com).

Congrats on the PR! It was quite delicate and it's working very well now :)

**Review by aaronfc — APPROVED** [review]

**Comment by wellyshen** [comment]
> I am not completely sure about this. I understand your point, but since this is an asynchronous process (sending an email). The users could login to WordPress.com before clicking the email link. And then they would land in WP.com.
Having said that, you are right. This might not be directly related to this PR, since I tested and it's already doing the same in production. Maybe is it worth a ticket to look into it?

I see, I have created a related issue for it: 105698-Automattic/gravatar
@aaronfc Thank you for the review and feedback 💯

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8871184

Thank you @wellyshen for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8871184).

**Pull request merged**